oscpy/server.py

```diff
@@ -308,69 +319,91 @@ def stop_all(self):
             self.stop(s)
         sleep(10e-9)
 
+    def terminate_server(self):
+        """Request the inner thread to finish its tasks and exit.
+
+        May be called from an event, too.
+        """
+        self._must_loop = False
+
+    def join_server(self, timeout=None):
+        """Wait for the server to exit (`terminate_server()` must have been called before).
+
+        Returns True iff the inner thread exited before timeout."""
+        return self._termination_event.wait(timeout=timeout)
+
     def _listen(self):
         """(internal) Busy loop to listen for events.
 
         This method is called in a thread by the `listen` method, and
         will be the one actually listening for messages on the server's
         sockets, and calling the callbacks when messages are received.
         """
-        match = self._match_address
-        advanced_matching = self.advanced_matching
-        addresses = self.addresses
-        stats = self.stats_received
-
-        while True:
-            drop_late = self.drop_late_bundles
-            if not self.sockets:
-                sleep(.01)
-                continue
-            else:
-                try:
-                    read, write, error = select(self.sockets, [], [], self.timeout)
-                except (ValueError, socket.error):
-                    continue
-
-            for sender_socket in read:
-                try:
-                    data, sender = sender_socket.recvfrom(65535)
-                except ConnectionResetError:
+        try:
```

Comment:
hm, not a fan of large indented sections if they can be avoided, maybe use a decorator to achieve that more nicely?

Reply:
A decorator could diminish clarity, maybe via a launcher method like in my new version?

Reply:
that's indeed a lot better :)